typing/typemod.ml

```diff
@@ -85,33 +87,112 @@ let extract_sig_open env loc mty =
 
 (* Compute the environment after opening a module *)
 
-let type_open_ ?used_slot ?toplevel ovf env loc lid =
-  let path = Typetexp.lookup_module ~load:true env lid.loc lid.txt in
-  match Env.open_signature ~loc ?used_slot ?toplevel ovf path env with
-  | Some env -> path, env
-  | None ->
-      let md = Env.find_module path env in
-      ignore (extract_sig_open env lid.loc md.md_type);
-      assert false
+let type_module_fwd : (Env.t -> Parsetree.module_expr ->
+                       Typedtree.module_expr) ref =
+  ref (fun _ _ -> assert false)
+
+let mod_ident_counter = ref 0
+let generated_module_ident = ref []
+
+let push_current_mid (mi, md, env) =
+  generated_module_ident := (mi, md, env) :: !generated_module_ident
+
+let pop_current_mid () =
+  let slot = List.hd !generated_module_ident in
+  generated_module_ident := List.tl !generated_module_ident;
+  slot
+
+let gen_mod_ident () =
+  let n = !mod_ident_counter in
+  incr mod_ident_counter;
+  let ident = Ident.create (Printf.sprintf "M#%d" n) in
```

Comment:
I find this naming scheme `M#%d` quite lacking since both the name `M` and the identifier `%d` are arbitrary and not really informative for users. Would it be possible to use `open#%location#optional_disambiguation_id` or a more expressive naming scheme?

Reply:
Also, cosmetic: `let ident = ... in ident` => `...`.